pkg/util/jsonmergepatch/json_merge_patch.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonmergepatch
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/evanphx/json-patch"
+)
+
+// Create a 3-way merge patch based-on JSON merge patch.
+// Calculate addition-and-change patch between current and modified.
+// Calculate deletion patch between original and modified.
+func CreateThreeWayJSONMergePatch(original, modified, current []byte, overwrite bool) ([]byte, error) {
+	addAndChangePatch, err := jsonpatch.CreateMergePatch(current, modified)
+	if err != nil {
+		return nil, err
+	}
+	// Only keep addition and changes
+	addAndChangePatch, err = keepOrDeleteNullInJsonPatch(addAndChangePatch, false)
+	if err != nil {
+		return nil, err
+	}
+
+	deletePatch, err := jsonpatch.CreateMergePatch(original, modified)
+	if err != nil {
+		return nil, err
+	}
+	// Only keep deletion
+	deletePatch, err = keepOrDeleteNullInJsonPatch(deletePatch, true)
```

Comment:
I used apply to create a TPR like this:
```
apiVersion: example.com/v1
kind: Widget
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"example.com/v1","kind":"Widget","metadata":{"annotations":{},"name":"foo","namespace":"default"},"spec":{"arbitraryData":{"foo":"bar"}}}
  creationTimestamp: 2017-01-23T21:10:43Z
  name: foo
  namespace: default
  resourceVersion: "1487"
  selfLink: /apis/example.com/v1/namespaces/default/widgets/foo
  uid: 66dd1772-e1b0-11e6-a97b-acbc32c1ca87
spec:
  arbitraryData:
    foo: bar
```

I then changed my local object so the spec part looked like this:
```
...
spec:
  arbitraryData:
    baz: qux
```

applying again computed this patch:
```
{
  "metadata":{
    "annotations":{
      "kubectl.kubernetes.io~1last-applied-configuration":"{\"apiVersion\":\"example.com/v1\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"foo\",\"namespace\":\"default\"},\"spec\":{\"arbitraryData\":{\"baz\":\"qux\"}}}\n"
    }
  },
  "spec":{
    "arbitraryData":{
      "baz":"qux"
    }
  }
}
```

it didn't send anything to the server indicating spec.arbitraryData.foo should be removed.